File: js/src/http.ts
/**
 * Internal `fetch` wrapper.
 *
 * Adds the `Authorization` header, JSON serializes the body, parses
 * the JSON response, and maps non-2xx envelopes (`{error: {code,
 * message, details}}`) into `NitropingError`. `fetch` itself failing
 * (DNS, TLS, offline, abort) becomes a `NetworkError`.
 *
 * Zero runtime deps — uses the platform's global `fetch`, `URL`, and
 * `AbortController`. Works in Node 18+, Bun, Deno, Cloudflare Workers
 * and modern browsers.
 */

import { NetworkError, NitropingError } from "./errors"

/** Default base URL pointing at the hosted nitroping service. */
export const DEFAULT_BASE_URL = "https://nitroping.dev"

/**
 * SDK version, surfaced in the default `User-Agent`. Keep in sync with
 * `package.json` (the repo-root `bump.sh` updates both on release).
 */
export const SDK_VERSION = "0.2.10"

/**
 * Constructor options shared between server and public-key clients.
 */
export interface HttpClientOptions {
  /**
   * Secret API key (`np_...`) or public key (`pk_...`). Sent in the
   * `Authorization` header. The scheme (`ApiKey` vs `Public`) is set
   * by `authScheme`.
   */
  apiKey: string
  /** Base URL. Defaults to `https://nitroping.dev`. */
  baseUrl?: string
  /**
   * Per-request timeout in milliseconds. Default: 30_000.
   *
   * Set to `0` to disable. Implemented with `AbortController` so it
   * works in every runtime.
   */
  timeoutMs?: number
  /** Custom `fetch` implementation. Useful for dependency injection in tests. */
  fetch?: typeof fetch
  /**
   * Authorization scheme. `ApiKey` for the server SDK, `Public` for
   * the browser-side public-key flow.
   */
  authScheme?: "ApiKey" | "Public"
  /**
   * Custom `User-Agent` header. Ignored in browsers (the runtime sets
   * it itself).
   */
  userAgent?: string
  /**
   * Debug logging. The SDK is silent by default. Set:
   *   * `true`  → log each request/response to `console.debug`.
   *   * a function → receive structured `{ phase, ... }` events to route
   *     anywhere (your own logger, Sentry breadcrumbs, etc.).
   * The `Authorization` header and the API key are always redacted.
   */
  debug?: boolean | DebugLogger
}

/** A structured debug event emitted by the HTTP client. */
export type DebugEvent =
  | {
      phase: "request"
      method: string
      url: string
      headers: Record<string, string>
      body?: unknown
    }
  | { phase: "response"; method: string; url: string; status: number; ms: number; body?: unknown }
  | { phase: "error"; method: string; url: string; ms: number; error: string }

/** Debug sink. Receives one event per request lifecycle step. */
export type DebugLogger = (event: DebugEvent) => void

/** Internal: a structured HTTP client. */
export class HttpClient {
  readonly baseUrl: string
  readonly apiKey: string
  readonly timeoutMs: number
  readonly fetchImpl: typeof fetch
  readonly authScheme: "ApiKey" | "Public"
  readonly userAgent: string
  private readonly debug?: DebugLogger

  constructor(opts: HttpClientOptions) {
    if (!opts.apiKey) {
      throw new NitropingError("apiKey is required", { code: "invalid_argument" })
    }
    this.apiKey = opts.apiKey
    this.baseUrl = (opts.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, "")
    this.timeoutMs = opts.timeoutMs ?? 30_000
    this.authScheme = opts.authScheme ?? (opts.apiKey.startsWith("pk_") ? "Public" : "ApiKey")
    this.userAgent = opts.userAgent ?? `nitroping-js/${SDK_VERSION}`
    this.debug = resolveDebug(opts.debug)

    const f = opts.fetch ?? globalThis.fetch
    if (typeof f !== "function") {
      throw new NitropingError(
        "Global `fetch` is not available. Pass a `fetch` implementation in the SDK constructor options.",
        { code: "fetch_unavailable" },
      )
    }
    this.fetchImpl = f.bind(globalThis)
  }

  /** Perform an HTTP request and parse the JSON envelope. */
  async request<T>(
    method: string,
    path: string,
    options: {
      body?: unknown
      headers?: Record<string, string>
      query?: Record<string, string | number | boolean | undefined>
    } = {},
  ): Promise<T> {
    const url = this.buildUrl(path, options.query)

    const headers: Record<string, string> = {
      Authorization: `${this.authScheme} ${this.apiKey}`,
      Accept: "application/json",
      ...options.headers,
    }

    let bodyText: string | undefined
    if (options.body !== undefined) {
      bodyText = JSON.stringify(options.body)
      headers["Content-Type"] = "application/json"
    }

    // Don't override the runtime's User-Agent in browsers (they reject
    // it). Only set it where the header is fully writable.
    if (typeof window === "undefined" && this.userAgent && !headers["User-Agent"]) {
      headers["User-Agent"] = this.userAgent
    }

    const controller = this.timeoutMs > 0 ? new AbortController() : undefined
    const timer =
      controller !== undefined ? setTimeout(() => controller.abort(), this.timeoutMs) : undefined

    const startedAt = nowMs()
    this.debug?.({
      phase: "request",
      method,
      url,
      headers: redactHeaders(headers),
      body: options.body,
    })

    let response: Response
    try {
      response = await this.fetchImpl(url, {
        method,
        headers,
        body: bodyText,
        signal: controller?.signal,
      })
    } catch (cause) {
      this.debug?.({
        phase: "error",
        method,
        url,
        ms: nowMs() - startedAt,
        error: (cause as Error)?.message ?? String(cause),
      })
      throw new NetworkError(
        `Request to ${url} failed: ${(cause as Error)?.message ?? cause}`,
        cause,
      )
    } finally {
      if (timer !== undefined) clearTimeout(timer)
    }

    try {
      const { value, raw } = await parseResponse<T>(response)
      this.debug?.({
        phase: "response",
        method,
        url,
        status: response.status,
        ms: nowMs() - startedAt,
        body: raw,
      })
      return value
    } catch (err) {
      // HTTP error envelope (NitropingError) — log it as a response phase so
      // debug users see the status + server code, then rethrow unchanged.
      this.debug?.({
        phase: "response",
        method,
        url,
        status: response.status,
        ms: nowMs() - startedAt,
        body: err instanceof NitropingError ? { code: err.code, message: err.message } : undefined,
      })
      throw err
    }
  }

  private buildUrl(
    path: string,
    query?: Record<string, string | number | boolean | undefined>,
  ): string {
    const url = new URL(path.startsWith("/") ? path : `/${path}`, `${this.baseUrl}/`)
    if (query) {
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined) url.searchParams.set(k, String(v))
      }
    }
    return url.toString()
  }
}

async function parseResponse<T>(response: Response): Promise<{ value: T; raw: unknown }> {
  const text = await response.text()
  let json: unknown = undefined
  if (text.length > 0) {
    try {
      json = JSON.parse(text)
    } catch {
      // Non-JSON body — pass through as raw text in the error.
      if (!response.ok) {
        throw new NitropingError(
          `HTTP ${response.status} ${response.statusText}: ${text.slice(0, 200)}`,
          { status: response.status, code: `http_${response.status}` },
        )
      }
      return { value: text as unknown as T, raw: text }
    }
  }

  if (!response.ok) {
    const envelope = (json ?? {}) as {
      error?: { code?: string; message?: string; details?: unknown }
    }
    const err = envelope.error ?? {}
    throw new NitropingError(err.message ?? `HTTP ${response.status}`, {
      status: response.status,
      code: err.code ?? `http_${response.status}`,
      details: err.details,
    })
  }

  return { value: json as T, raw: json }
}

/** High-resolution-ish monotonic ms, falling back to Date.now(). */
function nowMs(): number {
  return typeof performance !== "undefined" && typeof performance.now === "function"
    ? performance.now()
    : Date.now()
}

/** Strip the Authorization header before it ever reaches a log sink. */
function redactHeaders(headers: Record<string, string>): Record<string, string> {
  const out: Record<string, string> = {}
  for (const [k, v] of Object.entries(headers)) {
    out[k] = k.toLowerCase() === "authorization" ? "[redacted]" : v
  }
  return out
}

/** Normalize the `debug` option into a logger (or undefined when off). */
function resolveDebug(debug?: boolean | DebugLogger): DebugLogger | undefined {
  if (!debug) return undefined
  if (typeof debug === "function") return debug
  return (event) => {
    // Default sink: a single console.debug line per event.
    const label = `[nitroping] ${event.phase} ${event.method} ${event.url}`
    if (event.phase === "response") {
      console.debug(`${label} → ${event.status} (${Math.round(event.ms)}ms)`, event.body ?? "")
    } else if (event.phase === "error") {
      console.debug(`${label} ✗ ${event.error} (${Math.round(event.ms)}ms)`)
    } else {
      console.debug(label, event.body ?? "")
    }
  }
}

File: go/http.go
package nitroping

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// Version is the SDK's released version string. The Go module is
// versioned via git tags (there is no manifest file to read at build
// time), so this constant is the single source of truth for the
// User-Agent. Bump it in lockstep with each tagged release.
const Version = "0.2.12"

// defaultUserAgent is appended to every outbound request so server-side
// logs can distinguish SDK traffic from raw curl/Postman. Tracks
// `userAgent` in the JS SDK's http.ts.
const defaultUserAgent = "nitroping-go/" + Version

// transport is the package-internal HTTP layer shared by the
// notifications and devices resources. It owns the base URL, the API
// key, the http.Client implementation and the User-Agent. All resource
// methods route through transport.do so behaviour (auth, content-type,
// error envelope parsing) lives in one place.
type transport struct {
	baseURL    string
	apiKey     string
	authScheme string
	userAgent  string
	httpClient *http.Client
	// debug, when non-nil, receives a structured event map on each
	// request, response, and transport error. The Authorization header /
	// API key is always redacted before the event is emitted.
	debug func(event map[string]any)
}

// emitDebug forwards a structured event to the configured debug logger
// (if any). It is a no-op when debug is nil, which is the default.
func (t *transport) emitDebug(event map[string]any) {
	if t.debug == nil {
		return
	}
	t.debug(event)
}

// do issues an HTTP request to the nitroping API and decodes the JSON
// response into `out`. Non-2xx responses with a recognisable
// `{"error": {...}}` envelope are decoded into *APIError; transport-
// level failures (DNS, TLS, timeout, request cancellation) are wrapped
// with %w so callers can errors.Is them against context.Canceled etc.
//
// `out` may be nil if the caller doesn't care about the body (e.g.
// success-only 200 OK ack). When out is non-nil the response body must
// be valid JSON or a wrapping error is returned.
func (t *transport) do(
	ctx context.Context,
	method, path string,
	body any,
	cfg requestConfig,
	out any,
) error {
	endpoint, err := t.buildURL(path)
	if err != nil {
		return fmt.Errorf("nitroping: invalid URL %q: %w", path, err)
	}

	var reqBody io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("nitroping: marshal request body: %w", err)
		}
		reqBody = bytes.NewReader(buf)
	}

	req, err := http.NewRequestWithContext(ctx, method, endpoint, reqBody)
	if err != nil {
		return fmt.Errorf("nitroping: build request: %w", err)
	}

	req.Header.Set("Authorization", t.authScheme+" "+t.apiKey)
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if t.userAgent != "" {
		req.Header.Set("User-Agent", t.userAgent)
	}
	if cfg.idempotencyKey != "" {
		req.Header.Set("Idempotency-Key", cfg.idempotencyKey)
	}
	for k, v := range cfg.extraHeaders {
		// Don't let WithHeader override the auth / content-type /
		// accept headers we just set — those are integral to the
		// protocol.
		switch http.CanonicalHeaderKey(k) {
		case "Authorization", "Content-Type", "Accept", "User-Agent":
			continue
		}
		req.Header.Set(k, v)
	}

	t.emitDebug(map[string]any{
		"phase":  "request",
		"method": method,
		"url":    endpoint,
		// Authorization is intentionally not included — never log the
		// API key. We surface only the scheme so callers can tell which
		// auth mode was used without leaking the secret.
		"auth_scheme":   t.authScheme,
		"authorization": "[REDACTED]",
	})

	resp, err := t.httpClient.Do(req)
	if err != nil {
		t.emitDebug(map[string]any{
			"phase":  "error",
			"method": method,
			"url":    endpoint,
			"error":  err.Error(),
		})
		return fmt.Errorf("nitroping: %s %s: %w", method, endpoint, err)
	}
	defer resp.Body.Close()

	respBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		t.emitDebug(map[string]any{
			"phase":  "error",
			"method": method,
			"url":    endpoint,
			"status": resp.StatusCode,
			"error":  err.Error(),
		})
		return fmt.Errorf("nitroping: read response body: %w", err)
	}

	t.emitDebug(map[string]any{
		"phase":  "response",
		"method": method,
		"url":    endpoint,
		"status": resp.StatusCode,
	})

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return decodeAPIError(resp.StatusCode, respBytes)
	}

	if out == nil || len(respBytes) == 0 {
		return nil
	}
	if err := json.Unmarshal(respBytes, out); err != nil {
		return fmt.Errorf("nitroping: decode response body: %w", err)
	}
	return nil
}

// buildURL resolves `path` against the configured base URL. The base
// URL has had any trailing slashes stripped at client construction so
// we just glue with "/".
func (t *transport) buildURL(path string) (string, error) {
	if !strings.HasPrefix(path, "/") {
		path = "/" + path
	}
	full := t.baseURL + path
	// Validate the result; this catches stray spaces and bogus base
	// URLs early with a clear error.
	if _, err := url.Parse(full); err != nil {
		return "", err
	}
	return full, nil
}

// errorEnvelope mirrors the server's JSON error shape:
//
//	{"error": {"code": "...", "message": "...", "details": {...}}}
//
// `Details` is decoded into a generic map so callers can introspect
// per-field validation errors without a typed struct per error code.
type errorEnvelope struct {
	Error struct {
		Code    string         `json:"code"`
		Message string         `json:"message"`
		Details map[string]any `json:"details"`
	} `json:"error"`
}

// decodeAPIError turns a non-2xx response into a *APIError. It is
// tolerant: a missing envelope, bad JSON, or empty body all fall back
// to a synthetic message + a code of `http_<status>` so callers always
// get structured information they can branch on.
func decodeAPIError(status int, body []byte) error {
	apiErr := &APIError{StatusCode: status}

	if len(body) > 0 {
		var env errorEnvelope
		if err := json.Unmarshal(body, &env); err == nil && (env.Error.Code != "" || env.Error.Message != "") {
			apiErr.Code = env.Error.Code
			apiErr.Message = env.Error.Message
			apiErr.Details = env.Error.Details
		} else {
			// No structured envelope — include the raw body (truncated)
			// in the message so debugging isn't impossible.
			trimmed := body
			if len(trimmed) > 256 {
				trimmed = trimmed[:256]
			}
			apiErr.Message = strings.TrimSpace(string(trimmed))
		}
	}

	if apiErr.Code == "" {
		apiErr.Code = fmt.Sprintf("http_%d", status)
	}
	if apiErr.Message == "" {
		apiErr.Message = http.StatusText(status)
	}
	return apiErr
}

// sanityCheckBaseURL guards against the most common config mistake:
// passing a non-http(s) URL. Returns the trimmed URL or an error.
func sanityCheckBaseURL(raw string) (string, error) {
	trimmed := strings.TrimRight(raw, "/")
	if trimmed == "" {
		return "", errors.New("baseURL cannot be empty")
	}
	u, err := url.Parse(trimmed)
	if err != nil {
		return "", err
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return "", fmt.Errorf("baseURL scheme must be http or https, got %q", u.Scheme)
	}
	if u.Host == "" {
		return "", errors.New("baseURL must include a host")
	}
	return trimmed, nil
}
